cmd/thv-operator/controllers/mcpserver_runconfig.go

```diff
@@ -0,0 +1,456 @@
+	}
+
+	// ConfigMap exists, check if content has changed by comparing checksums
+	currentChecksum := current.Annotations["toolhive.stacklok.io/content-checksum"]
+	desiredChecksum := desired.Annotations["toolhive.stacklok.io/content-checksum"]
+
+	if currentChecksum != desiredChecksum {
+		// Content changed, update the ConfigMap with new checksum
+		// Copy resource version and other metadata for update
+		desired.ResourceVersion = current.ResourceVersion
+		desired.UID = current.UID
+
+		if err := controllerutil.SetControllerReference(mcpServer, desired, r.Scheme); err != nil {
+			return fmt.Errorf("failed to set controller reference for RunConfig ConfigMap: %w", err)
+		}
+
+		ctxLogger.Info("RunConfig ConfigMap content changed, updating",
+			"ConfigMap.Name", desired.Name,
+			"oldChecksum", currentChecksum,
+			"newChecksum", desiredChecksum)
+		if err := r.Update(ctx, desired); err != nil {
+			return fmt.Errorf("failed to update RunConfig ConfigMap: %w", err)
+		}
+		ctxLogger.Info("RunConfig ConfigMap updated", "ConfigMap.Name", desired.Name)
+	}
+
+	return nil
+}
+
+// runConfigContentEquals compares the actual content of RunConfig ConfigMaps using checksums
+func (*MCPServerReconciler) runConfigContentEquals(current, desired *corev1.ConfigMap) bool {
+	// Compare checksums - if both have checksums, use them for comparison
+	currentChecksum := current.Annotations["toolhive.stacklok.io/content-checksum"]
+	desiredChecksum := desired.Annotations["toolhive.stacklok.io/content-checksum"]
+
+	if currentChecksum != "" && desiredChecksum != "" {
+		return currentChecksum == desiredChecksum
+	}
+
+	// Fallback to compute checksums if they don't exist (for backward compatibility)
+	if currentChecksum == "" {
+		currentChecksum = computeConfigMapChecksum(current)
+	}
+	if desiredChecksum == "" {
+		desiredChecksum = computeConfigMapChecksum(desired)
+	}
+
+	return currentChecksum == desiredChecksum
+}
+
+// createRunConfigFromMCPServer converts MCPServer spec to RunConfig using the builder pattern
+// This creates a RunConfig for serialization to ConfigMap, not for direct execution
+func (*MCPServerReconciler) createRunConfigFromMCPServer(m *mcpv1alpha1.MCPServer) (*runner.RunConfig, error) {
+	proxyHost := defaultProxyHost
+	if envHost := os.Getenv("TOOLHIVE_PROXY_HOST"); envHost != "" {
+		proxyHost = envHost
+	}
+
+	port := 8080
+	if m.Spec.Port != 0 {
+		port = int(m.Spec.Port)
+	}
+
+	// Convert environment variables from MCPServer spec
+	envVars := make(map[string]string, len(m.Spec.Env))
+	for _, env := range m.Spec.Env {
+		envVars[env.Name] = env.Value
+	}
+
+	// Convert volumes from MCPServer spec
+	volumes := make([]string, 0, len(m.Spec.Volumes))
+	for _, vol := range m.Spec.Volumes {
+		volStr := fmt.Sprintf("%s:%s", vol.HostPath, vol.MountPath)
+		if vol.ReadOnly {
+			volStr += ":ro"
+		}
+		volumes = append(volumes, volStr)
+	}
+
+	// Convert secrets from MCPServer spec
+	secrets := make([]string, 0, len(m.Spec.Secrets))
```

Comment:
And these, similar to what we do with the `NewMCPServerPodTemplateSpecBuilder` below. This should hide some of those code away